sled-agent/src/zone_bundle.rs

```diff
@@ -6,42 +6,466 @@
+        out
+    }
+}
+
+impl ZoneBundler {
+    // A task run in the background that periodically cleans up bundles.
+    //
+    // This waits for:
+    //
+    // - A timeout at the current cleanup period
+    // - A notification that the cleanup context has changed.
+    //
+    // When needed, it actually runs the period cleanup itself, using the
+    // current context.
+    async fn periodic_cleanup(self) {
+        let (mut next_cleanup, mut time_to_next_cleanup) =
+            self.inner.lock().await.next_cleanup();
+        loop {
+            info!(
+                self.log,
+                "top of bundle cleanup loop";
+                "next_cleanup" => ?&next_cleanup,
+                "time_to_next_cleanup" => ?time_to_next_cleanup,
+            );
+
+            // Wait for the cleanup period to expire, or a notification that the
+            // context has been changed.
+            tokio::select! {
+                _ = sleep(time_to_next_cleanup) => {
+                    debug!(self.log, "running automatic periodic zone bundle cleanup");
+                    let mut inner = self.inner.lock().await;
+                    let dirs = inner.resources.all_zone_bundle_directories().await;
+                    let res = run_cleanup(&self.log, &dirs, &inner.cleanup_context).await;
+                    inner.last_cleanup_at = Instant::now();
+                    (next_cleanup, time_to_next_cleanup) = inner.next_cleanup();
+                    debug!(self.log, "cleanup completed"; "result" => ?res);
+                }
+                _ = self.notify_cleanup.notified() => {
+                    debug!(self.log, "notified about cleanup context change");
+                    let inner = self.inner.lock().await;
+                    (next_cleanup, time_to_next_cleanup) = inner.next_cleanup();
+                }
+            }
+        }
+    }
+
+    /// Create a new zone bundler.
     ///
-    /// Logs are periodically archived out of their original location, and onto
-    /// one or more U.2 drives. This field is used to specify that archive
-    /// location, so that rotated logs for the zone's services may be found.
-    pub extra_log_dirs: Vec<Utf8PathBuf>,
-    /// Any zone-specific commands that will be part of the zone bundle.
+    /// This creates an object that manages zone bundles on the system. It can
+    /// be used to create bundles from running zones, and runs a period task to
+    /// clean them up to free up space.
+    pub fn new(
+        log: Logger,
+        resources: StorageResources,
+        cleanup_context: CleanupContext,
+    ) -> Self {
+        let notify_cleanup = Arc::new(Notify::new());
+        let inner = Arc::new(Mutex::new(Inner {
+            resources,
+            cleanup_context,
+            last_cleanup_at: Instant::now(),
+        }));
+        let self_ = Self { log, inner, notify_cleanup };
+        let self_clone = self_.clone();
+        tokio::task::spawn(self_clone.periodic_cleanup());
```

Comment:
I don't know what the implications are of trying to make a singleton tokio task, especially when unit tests (which involves starting multiple tokio runtimes) are involved. Maybe just the first one gets to start the task, and when that runtime exits the singleton task that should've been running forever is cancelled? If we really want a singleton it probably ought to be a thread instead of a task, at which point a bunch of stuff might need to change.

I don't have a good suggestion here. I think in other places where we're expecting one and only one "thing", we start it up early (somewhere in the bootstrap-agent, probably?) and just keep the handle around forever, and rely on no one trying to instantiate another one. I.e., no safety here, just have to do the right thing? :-/

Reply:
> I think in other places where we're expecting one and only one "thing", we start it up early (somewhere in the bootstrap-agent, probably?) and just keep the handle around forever, and rely on no one trying to instantiate another one. I.e., no safety here, just have to do the right thing? :-/

Yeah, that's definitely the pattern that other parts of the sled agent have done as well. And yeah, not awesome.

But until we can think of a better pattern, I think we _should_ still move the task handle into the `Inner`, and call `.abort()` on it when the owner is dropped. I'll do that now.